cluster flag is not passed to valkey-benchmark

Note that "cluster_mode": True, in the parsed options:

```
# python benchmark.py --mode client --target-ip 192.168.1.1 --valkey-benchmark-path /usr/local/bin/valkey-benchmark --use-running-server
=== Processing commit: HEAD ===
2026-01-08 18:39:56,148 [INFO] Loaded config: {'duration': 120, 'keyspacelen': [10000000], 'data_sizes': [16], 'pipelines': [10], 'clients': [50], 'commands': ['SET', 'GET', 'RPUSH', 'LPUSH', 'LPOP', 'SADD', 'SPOP', 'HSET', 'MGET', 'LRANGE', 'ZPOPMIN'], 'cluster_mode': True, 'tls_mode': False, 'warmup': 10, 'benchmark-threads': 2}
2026-01-08 18:39:56,148 [INFO] Detected architecture: x86_64
2026-01-08 18:39:56,148 [INFO] Using pre-built Valkey instance.
2026-01-08 18:39:56,148 [INFO] Commit HEAD | TLS=off | Cluster=on
2026-01-08 18:39:56,148 [INFO] Running benchmark with io_threads=None
2026-01-08 18:39:56,148 [INFO] Using custom valkey-benchmark path: /usr/local/bin/valkey-benchmark
2026-01-08 18:39:56,148 [INFO] Waiting for Valkey server to be ready from the benchmark client...
...
2026-01-08 18:39:56,162 [INFO] Running: /usr/local/bin/valkey-benchmark -h 192.168.1.1 -p 6379 --duration 120 -r 10000000 -d 16 -P 10 -c 50 -t SET --threads 2 --warmup 10 --seed 857625 --csv
2026-01-08 18:39:56,234 [ERROR] Command failed with exit code 1: /usr/local/bin/valkey-benchmark -h 192.168.1.1 -p 6379 --duration 120 -r 10000000 -d 16 -P 10 -c 50 -t SET --threads 2 --warmup 10 --seed 857625 --csv
2026-01-08 18:39:56,234 [ERROR] Command stderr: Error from server: MOVED 2929 10.42.3.27:6379
Error from server: MOVED 14405 10.42.1.35:6379
```

Note that the flag `--cluster` was not passed to valkey-benchmark tool. When ran manually, it works:

```
# /usr/local/bin/valkey-benchmark -h 192.168.1.1 -p 6379 --duration 120 -r 10000000 -d 16 -P 10 -c 50 -t SET --threads 2 --warmup 10 --seed 857625 --csv --cluster
Cluster has 3 primary nodes:

Node 0(Primary): 196d0406ce690914a1e5f97bf44ee8494bcbf6a4 10.42.3.27:6379
Node 1(Primary): 43a4b3a698117c302685eae8e9914864fb20a3ea 10.42.0.34:6379
Node 2(Primary): 7540012a38f5f3295ff9d25123b1270ad6fe3235 10.42.1.35:6379

"test","rps","avg_latency_ms","min_latency_ms","p50_latency_ms","p95_latency_ms","p99_latency_ms","max_latency_ms"
"SET","76480.09","6.413","0.280","5.919","11.199","18.895","91.711"
```

--- a/tests/test_benchmark_command.py
+++ b/tests/test_benchmark_command.py
@@ -57,6 +57,41 @@ class TestBuildBenchmarkCommandSimpleFormat:
         )
         assert "taskset" not in cmd
 
+    def test_simple_format_includes_cluster_flag_in_cluster_mode(
+        self, minimal_client_runner, base_cmd_params
+    ):
+        """Simple format commands include --cluster when cluster mode is enabled."""
+        minimal_client_runner.cluster_mode = True
+
+        cmd = minimal_client_runner._build_benchmark_command(
+            command="SET", **base_cmd_params
+        )
+
+        assert "--cluster" in cmd
+
+    def test_simple_format_omits_cluster_flag_when_cluster_mode_disabled(
+        self, minimal_client_runner, base_cmd_params
+    ):
+        """Simple format commands should not include --cluster outside cluster mode."""
+        cmd = minimal_client_runner._build_benchmark_command(
+            command="SET", **base_cmd_params
+        )
+
+        assert "--cluster" not in cmd
+
+    def test_simple_format_includes_cluster_flag_without_cluster_nodes_config(
+        self, minimal_client_runner, base_cmd_params
+    ):
+        """Issue #27 regression: cluster mode should not depend on cluster_nodes metadata."""
+        minimal_client_runner.cluster_mode = True
+        minimal_client_runner.config.pop("cluster_nodes", None)
+
+        cmd = minimal_client_runner._build_benchmark_command(
+            command="SET", **base_cmd_params
+        )
+
+        assert "--cluster" in cmd
+
 
 class TestBuildBenchmarkCommandTLS:
     """Test TLS mode includes TLS flags."""
@@ -143,3 +178,68 @@ class TestBuildBenchmarkCommandDuration:
         assert "-n" in cmd
         assert cmd[cmd.index("-n") + 1] == "5000"
         assert "--duration" not in cmd
+
+
+class TestBuildBenchmarkCommandScenarios:
+    """Test scenario-based command construction."""
+
+    def test_single_node_scenario_includes_cluster_flag_in_cluster_mode(
+        self, minimal_client_runner
+    ):
+        """Scenario commands include --cluster for single-node execution in cluster mode."""
+        minimal_client_runner.cluster_mode = True
+
+        cmd = minimal_client_runner._build_benchmark_command(
+            scenario={
+                "command": "SET foo bar",
+                "type": "write",
+                "cluster_execution": "single",
+            }
+        )
+
+        assert "--cluster" in cmd
+
+    def test_single_node_scenario_includes_cluster_flag_without_cluster_nodes_config(
+        self, minimal_client_runner
+    ):
+        """Scenario cluster routing should not depend on cluster_nodes metadata."""
+        minimal_client_runner.cluster_mode = True
+        minimal_client_runner.config.pop("cluster_nodes", None)
+
+        cmd = minimal_client_runner._build_benchmark_command(
+            scenario={
+                "command": "SET foo bar",
+                "type": "write",
+                "cluster_execution": "single",
+            }
+        )
+
+        assert "--cluster" in cmd
+
+    def test_parallel_scenario_omits_cluster_flag(self, minimal_client_runner):
+        """Parallel cluster execution should not pass --cluster to a single command."""
+        minimal_client_runner.cluster_mode = True
+
+        cmd = minimal_client_runner._build_benchmark_command(
+            scenario={
+                "command": "SET foo bar",
+                "type": "write",
+                "cluster_execution": "parallel",
+            }
+        )
+
+        assert "--cluster" not in cmd
+
+    def test_single_node_scenario_omits_cluster_flag_when_cluster_mode_disabled(
+        self, minimal_client_runner
+    ):
+        """Scenario commands should not include --cluster outside cluster mode."""
+        cmd = minimal_client_runner._build_benchmark_command(
+            scenario={
+                "command": "SET foo bar",
+                "type": "write",
+                "cluster_execution": "single",
+            }
+        )
+
+        assert "--cluster" not in cmd

--- a/valkey_benchmark.py
+++ b/valkey_benchmark.py
@@ -229,6 +229,14 @@ class ClientRunner:
             return self.config["cluster_ports"]
         return [self.config.get("port", 6379)]
 
+    def _should_add_cluster_flag(self, scenario: Optional[dict] = None) -> bool:
+        """Return whether the valkey-benchmark command should include --cluster."""
+        if not self.cluster_mode:
+            return False
+        if scenario is None:
+            return True
+        return scenario.get("cluster_execution", "single") == "single"
+
     def _flush_database(self) -> None:
         """Flush all data from the database before benchmark runs."""
         logging.info(
@@ -649,9 +657,8 @@ class ClientRunner:
             if scenario.get("sequential", False):
                 cmd += ["--sequential"]
 
-            if scenario.get("cluster_execution") == "single":
-                if self.cluster_mode and self.config.get("cluster_nodes"):
-                    cmd += ["--cluster"]
+            if self._should_add_cluster_flag(scenario):
+                cmd += ["--cluster"]
 
             # Seed: Default ON unless explicitly disabled with "seed": false
             if (
@@ -685,6 +692,9 @@ class ClientRunner:
 
             if sequential:
                 cmd += ["--sequential"]
+
+            if self._should_add_cluster_flag():
+                cmd += ["--cluster"]
 
             # Unified seed logic: Default ON unless config disables
             if self.config.get("seed") is not False:

